src/Serilog.AspNetCore/AspNetCore/RequestLoggingMiddleware.cs

```diff
@@ -54,31 +56,58 @@ public async Task Invoke(HttpContext httpContext)
         if (httpContext == null) throw new ArgumentNullException(nameof(httpContext));
 
         var start = Stopwatch.GetTimestamp();
-
         var collector = _diagnosticContext.BeginCollection();
+        var logger = _logger;
+        var level = LogEventLevel.Information; // only used if _getLevel throws an exception
+        double elapsedMs;
+
         try
         {
             await _next(httpContext);
-            var elapsedMs = GetElapsedMilliseconds(start, Stopwatch.GetTimestamp());
+            elapsedMs = GetElapsedMilliseconds(start, Stopwatch.GetTimestamp());
             var statusCode = httpContext.Response.StatusCode;
-            LogCompletion(httpContext, collector, statusCode, elapsedMs, null);
+            logger = _logger ?? Log.ForContext<RequestLoggingMiddleware>();
+            level = _getLevel(httpContext, elapsedMs, null);
+            LogCompletion(httpContext, collector, statusCode, elapsedMs, logger, level, null);
         }
         catch (Exception ex)
             // Never caught, because `LogCompletion()` returns false. This ensures e.g. the developer exception page is still
             // shown, although it does also mean we see a duplicate "unhandled exception" event from ASP.NET Core.
-            when (LogCompletion(httpContext, collector, 500, GetElapsedMilliseconds(start, Stopwatch.GetTimestamp()), ex))
+            // The elapsedMs is required for calculating the level and for LogCompletion.
+            // The level is required in the finally block.
+            when ((elapsedMs = GetElapsedMilliseconds(start, Stopwatch.GetTimestamp())) >= 0
+                  && (level = _getLevel(httpContext, elapsedMs, ex)) >= 0
+                  && LogCompletion(httpContext, collector, 500, elapsedMs, logger, level, ex))
         {
         }
         finally
         {
+            await CallEnrichDiagnosticContextAsync(httpContext, logger, level);
```

Comment:
I may be off track, but won't this be called _after_ the `LogCompletion` calls in the `try` and `catch/when` blocks?